### DISCUSSION TITLE
Schema for point locations

### DISCUSSION THREAD

**Discussion by vinay0000** [eose-tools-org/eose-api#1] [General]
Need to define the object schema for a point location, which can be used to specify locations, grid-points, positions of entities such as Ground-stations, etc. 

in _TAT-C_  the schema is processed as [here](src/eose/tatc/point.py). 

_**Other related work:**_

_CZML_
[Position](https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/Position)
[PositionList](https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/PositionList)
[PositionListOfLists](https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/PositionListOfLists)

_SensorThings API has the following:_ 
[Location](https://docs.ogc.org/is/15-078r6/15-078r6.html#26)

_Related STK objects_
[Position](https://help.agi.com/stk/#../Subsystems/dataProviders/Content/html/dataProviders/Cartesian_Position~Ground_Location.htm?TocPath=How%2520to%2520Use%2520STK%257CReports%2520%2526%2520Graphs%257CData%2520Providers%2520Reference%257CAlphabetical%2520Listing%2520of%2520Data%2520Providers%257CCartesian%2520Position%257C_____1)
[All Position](https://help.agi.com/stk/#../Subsystems/dataProviders/Content/html/dataProviders/All_Position.htm?Highlight=position)
[All Position](https://help.agi.com/stk/#../Subsystems/dataProviders/Content/html/dataProviders/All_Position.htm?Highlight=all%20position)
[Coverage](https://help.agi.com/stk/#../Subsystems/dataProviders/Content/html/dataProviders/Grid_Point_Locations~Coverage.htm)
[J2000 Position](https://help.agi.com/stk/#../Subsystems/dataProviders/Content/html/dataProviders/J2000_Position.htm?TocPath=How%2520to%2520Use%2520STK%257CReports%2520%2526%2520Graphs%257CData%2520Providers%2520Reference%257CAlphabetical%2520Listing%2520of%2520Data%2520Providers%257C_____204)
[J2000_Cartesian_Position](https://help.agi.com/stk/#../Subsystems/dataProviders/Content/html/dataProviders/J2000_Cartesian_Position.htm?TocPath=How%2520to%2520Use%2520STK%257CReports%2520%2526%2520Graphs%257CData%2520Providers%2520Reference%257CAlphabetical%2520Listing%2520of%2520Data%2520Providers%257C_____203)

Questions:
- What all higher level objects would this Point Location object support? (grid-points, ground-stations, polygons, other facilities, spacecrafts?) 
- What all coordinate systems should be supported?
- What should be the key/value pairs, their datatypes, and associated value-range?

@ryanketzner @jsipps26

**Comment by ptgrogan** [comment]
Although not reflected in the TAT-C schema, we have found good utility with building on [Shapely Point](https://shapely.readthedocs.io/en/stable/reference/shapely.Point.html) which simply expresses a point in terms of its x-, y-, and optionally z-coordinates. The coordinates are interpreted in a Coordinate Reference System (CRS) such as WGS84 (latitude/longitude/altitude).

It is complicated to support different coordinate systems because there is a large variety. Two approaches: (1) choose and document a standard (e.g., WGS84) and require the user to transform for their end use case; (2) add a CRS string in EPSG format (e.g., WGS84 is EPSG:4326). Note: GeoJSON started with (2) but ultimately selected (1) for simplicity.

Shapely Point maps to a [GeoJSON Point](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.2) geometry. Furthermore, a [GeoJSON Feature](https://datatracker.ietf.org/doc/html/rfc7946#section-3.2) wraps a geometry (namely: Point) with user-defined properties. It is also possible to define Feature [foreign members](https://datatracker.ietf.org/doc/html/rfc7946#section-6.1) though it is not necessarily broadly supported.

TAT-C exclusively uses Point to describe zero-dimensional observation targets on the Earth surface. The only additional property we have used previously is an integer identifier. It might be helpful to allow an optional text-based identifier/name.

**Reply by ptgrogan** [reply]
Here is an idea that allows for easy transformation to/from GeoJSON features:
```
"""
Candidate SurfacePoint Object Schema
"""

from enum import Enum
from typing import Union, Tuple

from geojson_pydantic import Feature, Point
from pydantic import BaseModel, Field, StrictInt, StrictStr, StrictFloat

class Body(str, Enum):
    """
    Enumeration of valid planetary bodies.
    """
    EARTH = "earth"
    MOON = "moon"
    MARS = "mars"

class SurfacePoint(BaseModel):
    """
    Point on or above the surface of a planetary body (compatible with GeoJSON Feature).
    """
    id: Union[StrictInt, StrictStr, StrictFloat] = Field(None, description="Point identifier.")
    body: Body = Field(Body.EARTH, description="Body on or above which this point exists.")
    coordinates: Union[Tuple[float, float], Tuple[float, float, float]] = Field(
        ...,
        description="Longitude (degrees), latitude (degrees), [optional] altitude (meters)"
    )

    def as_feature(self) -> Feature:
        """
        Transform this SurfacePoint into a GeoJSON Feature.
        """
        return Feature(
            type = "Feature", 
            geometry = Point(type="Point", coordinates=self.coordinates), 
            properties = {"id": self.id, "body": self.body}
        )

    @classmethod
    def from_feature(cls, feature: Feature) -> 'SurfacePoint':
        """
        Transform a GeoJSON Feature into a SurfacePoint.
        """
        if not isinstance(feature.geometry, Point):
            raise ValueError("Feature geometry must be a point.")
        return SurfacePoint(
            id = feature.properties.get("id",None),
            body = feature.properties.get("body", Body.EARTH),
            coordinates = feature.geometry.coordinates
        )
```
and a brief test script:
```
p1 = SurfacePoint(type="Feature", coordinates=[0,0,0], id=1, body=Body.EARTH)
p2 = SurfacePoint.from_feature(
    Feature(
        type = "Feature",
        geometry = Point(type = "Point", coordinates = [1,1,0]),
        properties = {"id": 2, "body": Body.EARTH}
    )
)
print(p1.model_dump_json())
print(p2.model_dump_json())
print(p1.as_feature().model_dump_json())
print(p2.as_feature().model_dump_json())
from geojson_pydantic import FeatureCollection
fc = FeatureCollection(type="FeatureCollection", features=[p1.as_feature(), p2.as_feature()])
import geopandas as gpd
df = gpd.GeoDataFrame.from_features(fc)
print(df)
```

**Reply by ptgrogan** [reply]
Building on one of @vinay0000 's comments below, we could constrain coordinates to a fixed domain (e.g., -180 to 180 longitude, as is done currently in TAT-C attributes) or leave it ambiguous if we want to provide flexibility in interpretation. GeoJSON does not appear to constrain coordinate values but the documentation has longitude from -180 to 180 and provides some recommendations for handling anti-meridian wrapping, etc.

**Comment by vinay0000** [comment]
I have summarized a separate discussion with @ryanketzner @jsipps26 below:

Approaching this from an **applications point of view**, the possible scenarios in the ``PointLocation`` schema shall be required are to specify:

- a point location on a (Earth's) surface. e.g. to specify the location of the ground-station 
- a 'list' of sparsely populated locations. e.g., set of ground-stations,  set of sparsely distributed ground-sensors, set of volcanoes, etc
- 'list' of observation locations (operations plan of satellite)
- Grid of locations. This shall be a massive quantity (>1000) of point locations, and it could in either structured or unstructured. Structured means that there is a relationship between the grid points (regularly spaced, gaussian distribution, etc). For this application we may want to have a separate PointGrid class itself (like discussed [here](https://github.com/eose-tools-org/eose-api/discussions/2))
- locations of UAVs (which usually have a Cartesian position/velocity state description). One could also think of a generic (Cartesian)  object, which becomes a 'property' of the UAVState object.

_To support the above applications:_ 

One could have a basic class which is purely geometric. Here, the relevant features to distinguish would be the **coordinates**, **topology**, and **conventions**. (Similar to that employed in [CZML](https://github.com/AnalyticalGraphicsInc/czml-writer/wiki/Position) where there are separate objects for Cartesian, Cartographic, Position, Velocity)

For example, you could imagine at least four versions of a coordinates which would be relevant to coverage analysis:

- S2 spherical Coordinate: A point specified by latitude/longitude constrained to a (topologically) spherical surface, which would gain its full 3D coordinates only in reference to a particular body. In the typical/default case, this would be WGS 84 ellipsoid, but could also be a spherical earth model or a more elaborate setup if desired (like a polyhedral model). Some basic conventions would need to be decided (e.g., radians or degrees, bounds at -180 180 or 0-360). Here, the difference between, say, geodetic and geocentric coordinates would not be baked into the point class, but would rather be specified as a matter of interpretation by the classes which use the Point class.

- R3 spherical coordinates: A point specified by latitude/longitude, and either altitude or radius.  Radius is perhaps a bit more flexible, since the point would be specified relative to a coordinate origin rather than a reference surface, but the simplicity of altitude for the typical use case (Earth coverage) might be worth the tradeoff.

- R3 Cartesian coordinates: A typical 3D point in cartesian coordinates.

- R2 Cartesian coordinates: A typical 2D point in Cartesian coordinates. Perhaps useful for running coverage simulations over a small, local area.

Then, the remaining features (coordinate system, reference body, etc.) could be specified in the Grid/PointGroup class, or any other class which has the Point class as a data member. This makes sense, since it is hard to imagine a situation where you would want to put different types of points (with different coordinate systems or conventions) within the same Grid class, and even if you did it wouldn't be hard to do. Single responsibility principle. So, the grid class might also hold a ReferenceFrame object specifying attitude w.r.t the reference body, 


@ptgrogan It seems to me that ``GeoJSON`` also follows similar philosophy, but for a subset of the above proposed features (i.e. it allows only R3 spherical coordinates), and limits the coordinate system to be WGS84 geodetic (or rather to a common system throughout the application). And your proposal before was to extend the GeoJSON feature to bodies other than Earth. Is that right?

**Reply by ptgrogan** [reply]
For collections of points, I would recommend simply using a list.

I agree with the four main types of coordinates above. S2 and R2 are the same if we specify different coordinate reference systems (e.g., EPSG:4326 versus EPSG:4087). GeoJSON took the path of only allowing WGS84 as an interchange format to simplify common representation. As mentioned in discussion #2, any sort of option we add turns into a pseudo-requirement for implementers. Sometimes, specificity may be desirable, other times it is onerous.

If we desired the feature to natively distinguish Earth versus non-Earth points, then we would need some sort of discriminant attribute to identify the celestial body, which is out of scope of _Geo_ JSON. This is not an immediate need for TAT-C but could be important for other applications (SpaceNet). Going down this path may require working with ambiguous data in some cases, for example, I am not aware of as well-defined coordinate systems on other planetary bodies (though I imagine they do exist).

**Comment by ptgrogan** [comment]
I posted a branch with an initial TargetPoint object model: https://github.com/eose-tools-org/eose-api/tree/dev-points

**Comment by vinay0000** [comment]
Comments based on commit fa05f3129c7d2e50271385c55538916db5140f50

- The name of the source folder is ``eose``. Should we instead make it ``eose-api`` or so?

- Should we allow for python version 12? Currently the requirement is ">=3.8, <3.11"

- In the ``__init__.py`` file, I had to change the line ``version = "0.0.1"`` to ``__version__ = "0.0.1"`` to make the automatic installation work.

-  I see the line ``bbox: Optional[BoundingBox] = None`` in the ``eose-api`` redefined classes (``Point``, ''LineString``, etc). Is this necessary or only for the sake of clarity? (I realize that GeoJSON allows for optional bounding box attribute.)

- The function ``targets.TargetPoint.as_targets`` would be executed everytime targets are required as a list of points. It is not efficient. Should we consider another (neat) alternate way. (Not a major issue.) 

- The ``id``s of the generated points in the grid (``grids.py`` module) are not going to be unique across multiple generated grids. This should be documented.
